Enhance ModelReaderWriter with collection support

m-nash opened the pull request:
Fixes https://github.com/Azure/azure-sdk-for-net/issues/48293

Updated ModelReaderWriter.cs to improve reading and writing functionality for various collection types, including List<> and Dictionary<>. Added error handling for unsupported types and enhanced exception messages.

Updated ModelReaderWriterTests.cs with new test cases to validate the changes, covering edge cases and empty collections. Added JSON test data for comprehensive testing of availability set data.

Included a new package reference for System.Memory.Data in project files. Minor corrections made to comments and variable names for clarity.

Revival of this old PR https://github.com/Azure/azure-sdk-for-net/pull/40598

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

m-nash committed:
Enhance ModelReaderWriter with collection support

Updated ModelReaderWriter.cs to improve reading and writing functionality for various collection types, including List<> and Dictionary<>. Added error handling for unsupported types and enhanced exception messages.

Updated ModelReaderWriterTests.cs with new test cases to validate the changes, covering edge cases and empty collections. Added JSON test data for comprehensive testing of availability set data.

Included a new package reference for System.Memory.Data in project files. Minor corrections made to comments and variable names for clarity.

JoshLove-msft commented on sdk/core/System.ClientModel/tests/ModelReaderWriter/Models/AvailabilitySetDataListOfListsTests.cs:
```diff
@@ -0,0 +1,239 @@
+
+        private static readonly HashSet<Type> s_supportedCollectionTypes =
+        [
+            typeof(List<>),
```
Should this use `IList<>` and `IDictionary<>`?

m-nash replied:
we need concrete constructable types not interfaces here

m-nash replied:
the restriction no longer exists so all types works

JoshLove-msft commented on sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs:
```diff
@@ -12,6 +15,12 @@ namespace System.ClientModel.Primitives;
 {
+    private static readonly HashSet<Type> s_supportedCollectionTypes =
+    [
+        typeof(List<>),
```
Should this use IList<> and IDictionary<>?

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[System.ClientModel](https://apiview.dev/Assemblies/Review/1b123e7a51d44ebe945f0212ee039c65?revisionId=e37882e1c8aa4a449a32d62a157b9b2c)

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into mrwLists

m-nash committed:
expand test coverage

m-nash committed:
update api

christothes commented on sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs:
```diff
@@ -12,6 +15,12 @@ namespace System.ClientModel.Primitives;
 /// </summary>
 public static class ModelReaderWriter
 {
+    private static readonly HashSet<Type> s_supportedCollectionTypes =
```
Do we expect the supported types to grow over time? With just a few elements an iterative lookup can be faster than the hash.

m-nash replied:
I think it would be easy to add things like Collection, SortedList, etc if we wanted to and this just centralized the checking to make it simpler than a series of if/else statements.

Still in discussion on which types to support for output

tg-msft replied:
How does this work for `ReadOnlyCollection<T>` (or whatever type we instantiate for an `IReadOnlyList<T>` output collection)?  Do we need to update this with every possible collection type we return?  How does it work for internal shared source types like `ChangeTrackingList<T>`?

m-nash replied:
For output the idea is we are limiting you to a set collection types.  Right now if you wanted a ChangeTrackingList<T> you would do this

```c#
ChangeTrackingList<MyType> list = new(ModelReaderWriter.Read<List<MyType>>(json));
```

For ReadOnlycollection<T> you would do
```c#
ReadOnlycollection<MyType> readonlyList = ModelReaderWriter.Read<List<MyType>>(json).AsReadOnly();
```

We can add support to do this internally, but we would need to do something special for each of these that we want to add so its a balance of how many types we think should be in this supported list.

m-nash replied:
the restriction no longer exists

m-nash committed:
update dotnet in ci

m-nash committed:
add aot annotations

m-nash committed:
more aot annotation

m-nash committed:
Update to keep current behavior but add new read methods which are AOT compatible for IEnumerable and IPersistableModel

m-nash committed:
fix aot on collection reader

m-nash committed:
update to remove collection read type restriction

m-nash committed:
update context pattern

m-nash committed:
drop internal interface

m-nash committed:
add ROM support

m-nash committed:
add immutablelist test example

m-nash committed:
add more test cases

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into mrwLists

m-nash committed:
code clean up

m-nash committed:
add coverage for context and non context into modeltests

m-nash committed:
refactor collection tests to use common base

m-nash committed:
update api

m-nash committed:
merge main

m-nash committed:
move list of list tests to inherit test base

m-nash committed:
update remaining tests

m-nash committed:
more test coverage

m-nash committed:
add remaining immutable and readonly tests

m-nash committed:
add perf tests to verify no regression and context path is faster

m-nash committed:
fix net framework test diff

m-nash committed:
add polymorphic collection tests

m-nash committed:
t status
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into mrwLists

weshaggard reviewed (approved):
eng changes look good.

m-nash committed:
comment out problematic tests

KrzysztofCwalina commented on sdk/core/System.ClientModel/src/ModelReaderWriter/JsonCollectionWriter.cs:
```diff
@@ -0,0 +1,90 @@
+        }
+    }
+
+    private static void WriteMultiDimensionalArray(Array array, int[] indices, int currentDimension, ModelReaderWriterOptions options, Utf8JsonWriter writer)
```
should we just throw for multi-dim arrays?

m-nash replied:
We could throw, but this does work today we have test cases for both [,] and [][] and it creates the write output in both cases.

m-nash commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,75 @@
+            {
+                return new JsonCollectionWriter();
+            }
+            throw new InvalidOperationException($"{persistableModel.GetType().Name} has a wire format of '{wireFormat}' it must be 'J' to be written as a collection");
```
We can add more CollectionWriters later, but I have only written the JsonCollectionWriter in this PR.  Adding XmlCollectionWriter should be fairly trivial as a follow up and that is why I made the CollectionWriter abstraction in the first place to be able to extend it later on.  Main downside to this approach is a 3rd party can't add their own collectionWriter until we make the base public and provide a way to pass theirs into the options class.

m-nash commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,75 @@
+                    return value;
+                }
+            }
+            throw new InvalidOperationException($"Can't use format 'W' format on an empty collection please specify a concrete format");
```
This code is choosing which CollectionWriter to invoke.  If we new we were supposed to choose the JsonCollectionWriter we could simply write "[]" which is exactly what we do if you used "J" format.

Without at least one item in the collection we can't call GetFormatForOptions to determine what the concrete format for 'W' is.

m-nash commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionReader.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        if (options.Format != "J" && options.Format != "W")
+        {
+            throw new InvalidOperationException($"Format '{options.Format}' is not supported only 'J' or 'W' format can be read as collections");
```
I haven't given this a ton of deep thought but was thinking along these same lines for some time.

There are 2 concepts here

1. Can I add my own custom formats for a given IPersistableModel<T>
2. Can I add my own custom collection scaffolding format

For 1 we could use the proxy work we started to take over the reading and writing for unknown formats by the model.
- check proxy format definitions
- check model format definitions
- throw format exception

For 2 it would be something along the lines of [this comment here](https://github.com/Azure/azure-sdk-for-net/pull/48102#discussion_r1984143657) .

JoshLove-msft commented on sdk/core/System.ClientModel/api/System.ClientModel.net8.0.cs:
```diff
@@ -223,13 +233,28 @@ public MessageLoggingPolicy(System.ClientModel.Primitives.ClientLoggingOptions?
         public sealed override void Process(System.ClientModel.Primitives.PipelineMessage message, System.Collections.Generic.IReadOnlyList<System.ClientModel.Primitives.PipelinePolicy> pipeline, int currentIndex) { }
         public sealed override System.Threading.Tasks.ValueTask ProcessAsync(System.ClientModel.Primitives.PipelineMessage message, System.Collections.Generic.IReadOnlyList<System.ClientModel.Primitives.PipelinePolicy> pipeline, int currentIndex) { throw null; }
     }
+    public abstract partial class ModelInfo
```
We should probably come up with a better name for this. Maybe ModelInitializer?

m-nash replied:
PersistableModelInfo?

JoshLove-msft replied:
My issue was more with "Info".

m-nash replied:
renamed to ModelBuilder per review fb

JoshLove-msft commented on sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs:
```diff
@@ -91,32 +157,103 @@ public static BinaryData Write(object model, ModelReaderWriterOptions? options =
+    }
+
+    private static T? ReadInternal<T>(
+    BinaryData data,
```
nit: indentation

JoshLove-msft commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+    /// <summary>
+    /// Gets the builder instance of the collection type.
+    /// </summary>
+    protected internal abstract object GetBuilder();
```
It's a little strange that the method is called GetBuilder and the type is also called CollectionBuilder. It would be nice to differentiate between the two somehow.

m-nash replied:
Resolved with the refactor on CollectionBuilder to make it internal

JoshLove-msft commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,75 @@
+    {
+        if (options.Format != "J" && options.Format != "W")
+        {
+            throw new InvalidOperationException($"Format '{options.Format}' is not supported only 'J' or 'W' format can be written as collections");
```
```suggestion
            throw new InvalidOperationException($"Format '{options.Format}' is not supported. Only 'J' or 'W' format can be written as collections.");
```

JoshLove-msft commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,75 @@
+            {
+                return new JsonCollectionWriter();
+            }
+            throw new InvalidOperationException($"{persistableModel.GetType().Name} has a wire format of '{wireFormat}' it must be 'J' to be written as a collection");
```
```suggestion
            throw new InvalidOperationException($"{persistableModel.GetType().Name} has a wire format of '{wireFormat}'. It must be 'J' to be written as a collection.");
```

JoshLove-msft commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+        else
+        {
+            throw new InvalidOperationException($"Unable to write {enumerable.GetType().Name} can only write collections of IPersistableModel");
```
```suggestion
            throw new InvalidOperationException($"Unable to write {enumerable.GetType().Name}. Only collections of 'IPersistableModel' can be written.");
```

azure-sdk commented:
**API Change Check**

jsquire commented on sdk/ai/Azure.AI.Projects/tests/AgentClientTests.cs:
```diff
@@ -49,8 +49,10 @@ public enum VecrorStoreTestType
         [RecordedTest]
-        [TestCase(ArgumentType.Metadata)]
-        [TestCase(ArgumentType.Bytes)]
+        //Failing in CI due to no playback found
```
Can we please open an issue to track or just rerecord the tests?

m-nash replied:
https://github.com/Azure/azure-sdk-for-net/issues/48725

jsquire commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+    /// </summary>
+    /// <param name="item">The item to add.</param>
+    /// <param name="key">Optional key for <see cref="IDictionary"/> collections.</param>
+    /// <exception cref="Exception"></exception>
```
nit:  Fill out or remove.  This just makes me a sad panda.

jsquire commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionReader.cs:
```diff
@@ -0,0 +1,36 @@
+            throw new InvalidOperationException($"Format '{options.Format}' is not supported only 'J' or 'W' format can be read as collections");
+        }
+
+        if (options.Format == "J")
```
It would be good to throw a comment in these explaining what "J" and "W" are.  Its not obvious for those unfamiliar with MRW.

jsquire commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,75 @@
+                    return value;
+                }
+            }
+            throw new InvalidOperationException($"Can't use format 'W' format on an empty collection please specify a concrete format");
```
```suggestion
            throw new InvalidOperationException($"Can't use format 'W' format on an empty collection. Please specify a concrete format");
```

jsquire replied:
I'm not in love with this message.  I wouldn't equate "a concrete format" with "go add something to this collection."

jsquire commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,75 @@
+            return enumerator.Current;
+        }
+
+        throw new InvalidOperationException($"Can't use format 'W' format on an empty collection please specify a concrete format");
```
Same here.

jsquire commented on sdk/core/System.ClientModel/src/ModelReaderWriter/JsonCollectionWriter.cs:
```diff
@@ -0,0 +1,90 @@
+        }
+        else
+        {
+            throw new InvalidOperationException($"{model.GetType().Name} does not implement IJsonModel or IEnumerable<IJsonModel>");
```
I assume that its ok, but are we sure `GetType()` is AOT friendly?

m-nash committed:
pr fb

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

eerhardt commented on sdk/core/System.ClientModel/src/ModelReaderWriter/ReflectionModelInfo.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace System.ClientModel.Primitives;
+
+internal class ReflectionModelInfo : ModelInfo
+{
+    private Type _type;
+
+    public ReflectionModelInfo(Type type)
+    {
+        _type = type;
+    }
+
+    //This will be replaced with RequiresUnreferencedCode when this issue is fixed https://github.com/Azure/azure-sdk-for-net/issues/48294
+    [UnconditionalSuppressMessage("Trimming", "IL2077",
+      Justification = "We will only call this when we went through ModelReaderWriter.Read which has DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors).")]
+    public override object CreateObject()
```
```suggestion
[RequiresUnreferencedCode("Uses reflection that can't be statically analyzed"]
internal class ReflectionModelInfo : ModelInfo
{
    private Type _type;

    public ReflectionModelInfo(Type type)
    {
        _type = type;
    }

    [UnconditionalSuppressMessage("Trimming", "IL2077",
      Justification = "This class is marked as RequiresUnreferencedCode.")]
    public override object CreateObject()
```

It's better to be more explicit that this whole class is not trim compatible. That way any callers are aware they need to reconsider calling this class in trimmed scenarios.

eerhardt replied:
Alternatively, if you can flow the `DynamicallyAccessedMembers` attribute to the constructor (and through `ModelReaderWriterContext.GetModelInfo`, it may work to be able to do this without a suppression.

eerhardt commented on sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs:
```diff
@@ -91,32 +157,103 @@ public static BinaryData Write(object model, ModelReaderWriterOptions? options =
+        return ReadInternal(
+            data,
+            returnType,
+            s_reflectionContext.Value,
```
This method needs to be marked `RequiresUnreferencedCode` because it uses a `ReflectionContext`, which requires unreferenced code.

m-nash commented on sdk/core/Azure.Core/api/Azure.Core.net6.0.cs:
```diff
@@ -17,12 +17,9 @@ protected AsyncPageable(System.Threading.CancellationToken cancellationToken) {
     }
     public static partial class AzureCoreExtensions
     {
-        [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCodeAttribute("This utilizes reflection-based JSON serialization and deserialization which is not compatible with trimming.")]
```
This is a visual bug only due to a tool issue.  I have opened this issue to track fixing the tool https://github.com/Azure/azure-sdk-for-net/issues/48753.

From a backwards compatibility check perspective it is already broken due to this issue which I filed on the apicompat tool https://github.com/dotnet/sdk/issues/47513

jsquire reviewed (commented):
General structure and format look good to me.

m-nash committed:
pr feedback to remove CollectionBuilder as a public class

m-nash committed:
more api feedback

m-nash committed:
fix param name

m-nash committed:
drop property funcs

m-nash committed:
drop assert key

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into mrwLists

m-nash committed:
drop asserts and createelementinstance

m-nash committed:
add support and test cases for null entries in the collections
add tests for custom collection

m-nash committed:
add support and tests for non generic collection and non string keys

m-nash committed:
Fill out coverage after refactors

m-nash committed:
update method names to match new renames

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into mrwLists

scbedd commented:
/azp run net - pullrequest

scbedd commented:
@m-nash any of the remaining failing checks will be cleared out by a new SHA. The latest run of [pullrequest](https://dev.azure.com/azure-sdk/public/_build/results?buildId=4664535&view=results) and [core](https://dev.azure.com/azure-sdk/public/_build/results?buildId=4659915&view=results) pipelines are both successful 👍

christothes commented on sdk/core/System.ClientModel/src/ModelReaderWriter/CollectionWriter.cs:
```diff
@@ -0,0 +1,76 @@
+
+    private static object GetFirstObject(IEnumerable enumerable)
+    {
+        if (enumerable is IDictionary dictionary)
```
Given a `Dictionary<foo, object>`, where some of the objects are `IEnumerable` and some are not, what would happen?

m-nash replied:
Added a test to show how you can make a mix work.  It requires you to write your own IJsonModel discriminator.  Writing mixed works with no custom discriminator but reading only works with.

m-nash committed:
added test cases for mixed model and enumerable

m-nash committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into mrwLists

KrzysztofCwalina reviewed (approved):
APIs look good to me
